src/App.tsx

```diff
@@ -1,61 +1,133 @@
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+import React, { useState } from 'react';
+import { User } from './types/User';
+
+function getUserById(userId: number): User | null {
+  return usersFromServer.find(person => person.id === userId) || null;
+}
+
+const todos: Todo[] = todosFromServer.map(todo => {
+  const user = getUserById(todo.userId);
+
+  return { ...todo, user };
+});
+
+function getNewTodoId(items: Todo[]) {
+  const maxId = Math.max(...items.map(item => item.id));
+
+  return maxId + 1;
+}
 
 export const App = () => {
+  const [visibleTodos, setVisibleTodos] = useState<Todo[]>(todos);
+  const [title, setTitle] = useState<string>('');
+  const [hasTitleError, setHasTitleError] = useState<boolean>(false);
+  const [selectedUser, setSelectedUser] = useState<number>(0);
+  const [hasSelectError, setHasSelectError] = useState<boolean>(false);
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setHasTitleError(false);
+  };
+
+  const handleSelectedUserChange = (
+    event: React.ChangeEvent<HTMLSelectElement>,
+  ) => {
+    setSelectedUser(+event.target.value);
+    setHasSelectError(false);
+  };
+
+  const onAdd = (newTodo: Todo) => {
+    const currentTodo = {
+      ...newTodo,
+      id: getNewTodoId(todos),
+    };
+
+    setVisibleTodos(currentTodos => [...currentTodos, currentTodo]);
+  };
+
+  const reset = () => {
+    setTitle('');
+    setSelectedUser(0);
+
+    setHasTitleError(false);
+    setHasSelectError(false);
+  };
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    setHasTitleError(!title);
+    setHasSelectError(!selectedUser);
```

Comment:
The error handling here sets the error states correctly, but consider providing more user feedback or logging to help users understand what went wrong.